File: schemas/content-moderation/mongodb/violation_categories.js
// violation_categories: Hierarchical taxonomy of content violation types.
// See README.md for full design rationale.

const mongoose = require("mongoose");

const violationCategorySchema = new mongoose.Schema(
  {
    name: { type: String, unique: true, required: true }, // Machine-readable identifier (e.g., "hate_speech", "csam").
    display_name: { type: String, required: true }, // Human-readable label (e.g., "Hate Speech").
    description: { type: String, default: null }, // Detailed explanation of what this category covers.
    parent_id: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "ViolationCategory",
      default: null,
    }, // Parent category for hierarchical taxonomy. Null = top-level.

    // info = informational/advisory.
    // low = minor policy violation.
    // medium = standard violation.
    // high = serious violation requiring prompt action.
    // critical = illegal content, imminent harm — highest priority.
    severity: {
      type: String,
      enum: ["info", "low", "medium", "high", "critical"],
      required: true,
      default: "medium",
    },

    is_active: { type: Boolean, required: true, default: true }, // Soft-disable without deleting.
    sort_order: { type: Number, required: true, default: 0 }, // Display ordering within the parent group.
  },
  {
    timestamps: { createdAt: "created_at", updatedAt: "updated_at" },
  }
);

// Indexes
violationCategorySchema.index({ parent_id: 1 });
violationCategorySchema.index({ is_active: 1, sort_order: 1 });

module.exports = mongoose.model("ViolationCategory", violationCategorySchema);

File: schemas/content-moderation/spacetimedb/violation_categories.rs
// violation_categories: Hierarchical taxonomy of content violation types.
// See README.md for full design rationale.

use spacetimedb::Timestamp;

/// info = informational/advisory.
/// low = minor policy violation.
/// medium = standard violation.
/// high = serious violation requiring prompt action.
/// critical = illegal content, imminent harm — highest priority.
// type: String
#[derive(SpacetimeType, Clone)]
pub enum ViolationCategorySeverity {
    Info,
    Low,
    Medium,
    High,
    Critical,
}

#[spacetimedb::table(name = violation_categories, public)]
pub struct ViolationCategory {
    #[primary_key]
    pub id: String, // UUID

    #[unique]
    pub name: String, // Machine-readable identifier (e.g., "hate_speech", "csam").

    pub display_name: String, // Human-readable label (e.g., "Hate Speech", "Child Sexual Abuse Material").

    pub description: Option<String>, // Detailed explanation of what this category covers.

    #[index(btree)]
    pub parent_id: Option<String>, // UUID — FK → violation_categories.id (restrict delete)
    // Null = top-level category.

    pub severity: ViolationCategorySeverity, // Default severity when this category is cited in an action.

    #[index(btree)]
    pub is_active: bool, // Soft-disable without deleting. Inactive categories cannot be selected for new violations but remain for history.

    pub sort_order: i32, // Display ordering within the parent group.
    // Composite index: (is_active, sort_order) — enforce in reducer logic

    pub created_at: Timestamp,
    pub updated_at: Timestamp,
}
